src/main/java/org/apache/sling/resourcemerger/impl/MergedResourcePickerWhiteboard.java

```diff
@@ -24,102 +24,96 @@
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
-import org.apache.felix.scr.annotations.Activate;
-import org.apache.felix.scr.annotations.Component;
-import org.apache.felix.scr.annotations.Deactivate;
 import org.apache.sling.api.resource.Resource;
 import org.apache.sling.api.resource.ResourceResolver;
 import org.apache.sling.commons.osgi.PropertiesUtil;
 import org.apache.sling.resourcemerger.spi.MergedResourcePicker;
 import org.apache.sling.resourcemerger.spi.MergedResourcePicker2;
 import org.apache.sling.spi.resource.provider.ResourceProvider;
 import org.osgi.framework.BundleContext;
-import org.osgi.framework.Constants;
-import org.osgi.framework.InvalidSyntaxException;
-import org.osgi.framework.ServiceReference;
 import org.osgi.framework.ServiceRegistration;
-import org.osgi.util.tracker.ServiceTracker;
-import org.osgi.util.tracker.ServiceTrackerCustomizer;
-
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * Registers all {@link MergedResourcePicker} and {@link MergedResourcePicker2} services as {@link MergingResourceProvider}s.
+ */
 @Component
-public class MergedResourcePickerWhiteboard implements ServiceTrackerCustomizer {
-
-    private ServiceTracker tracker;
+public class MergedResourcePickerWhiteboard {
 
+    
     private BundleContext bundleContext;
 
-    private final Map<Long, ServiceRegistration> serviceRegistrations = new ConcurrentHashMap<Long, ServiceRegistration>();
+    private final Map<String, ServiceRegistration<ResourceProvider>> resourceProvidersPerRoot = new ConcurrentHashMap<>();
```

Comment:
I guess I'm not sure if the change from using the service id of the used service as a key to using the merge root properties as a key is save. If you know for a fact that there are no two services coming along with the same merge root I guess it should be fine....